### PR TITLE
task 4

### PR DESCRIPTION
**Pull request by Jay7806**

### CODE REVIEW

File: controllers/data_controllers.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  selectTopics()
+    .then((topics) => {
+      res.status(200).send({ topics });
+    })
+    .catch((err) => {
+      next(err);
+    });
+};
+exports.getArticlesById = (req, res, next) => {
+  const { article_id } = req.query;
```
Comment (MRombi) [inline controllers/data_controllers.js]: we want to use req.params 🙂

File: __tests__/index.test.js [hunk 2]
```diff
@@ -64,8 +64,36 @@ describe("app", () => {
...
+      .expect(200)
+      .then((response) => {
+        expect(response.body.data).toEqual(endpointData);
+      });
+  });
+});
+describe("api/articles", () => {
+  test("200: Returns an article by the ID that is passed", () => {
+    return request(app)
+      .get("/api/articles?article_id=5")
```
Comment (MRombi) [inline __tests__/index.test.js]: This should look like this  .get("/api/articles/5")

File: db/seeds/app.js [hunk 3]
```diff
@@ -17,4 +18,19 @@ app.get("/api/topics", getTopics);
 
 app.get("/api", getApi);
 
+app.get("/api/articles", getArticlesById);
```
Comment (MRombi) [inline db/seeds/app.js]: This is the wrong endpoint, please look into parameters. This should look like this app.get("/api/articles/:article_id", getArticlesById); Let me know if it is not clear, but please look into the notes or the sprints from last week first 🙂

**Review by MRombi — CHANGES_REQUESTED** [review]
Good attempt on this ticket but please check the feedback I left and resubmit once you applied the changes 🙂

File: __tests__/index.test.js [hunk 4]
```diff
@@ -64,8 +64,36 @@ describe("app", () => {
...
+        expect(Array.isArray(articles)).toBe(true);
+        expect(articles).toHaveLength(1);
+        articles.forEach(({ article_id }) => {
+          expect(article_id).toBe(5);
+        });
+      });
+  });
+  test("404: receive a 404 error when article id is not found", () => {
+    return request(app)
+      .get("/api/articles?article_id=22")
```
Comment (johnhutcheon) [inline __tests__/index.test.js]: This line should also be changed to the same format as line 78 🙂

File: __tests__/index.test.js [hunk 5]
```diff
@@ -64,8 +64,36 @@ describe("app", () => {
...
+      .expect(404)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toEqual("Not found");
+      });
+  });
```
Comment (johnhutcheon) [inline __tests__/index.test.js:96]: We should also have a 400 error test here for if we send something that isn't a valid ID. I.e. api/articles/banana 👍

File: controllers/data_controllers.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    .catch((err) => {
+      next(err);
+    });
+};
+exports.getArticlesById = (req, res, next) => {
+  const { article_id } = req.params;
+  selectArticleById(article_id)
+    .then((articles) => {
+      if (articles.length === 0) {
+        res.status(404).send({ msg: "Not found" });
```
Comment (johnhutcheon) [inline controllers/data_controllers.js]: This 404 is already being handled in your model, so can be removed in here 👍

File: db/seeds/app.js [hunk 7]
```diff
@@ -17,4 +18,19 @@ app.get("/api/topics", getTopics);
...
+app.get("/api/articles/:article_id", getArticlesById);
+
+app.use((req, res) => {
+    res.status(404).send({msg: 'Not found'})
+});
+app.use((err, req, res, next) => {
+  if (err.status) {
+    res.status(err.status).json({ msg: err.msg });
+  } else {
+    console.error(err);
```
Comment (johnhutcheon) [inline db/seeds/app.js]: Ensure you're removing any console logs before submitting code

**Review by johnhutcheon — CHANGES_REQUESTED** [review]
Hey James, thanks for making the changes Michael asked for. Could you just see my comments and make the further changes please? Thanks! 🙂

**Pull request merged**